draft-irtf-cfrg-opaque.md

```diff
@@ -393,152 +397,342 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
 We let `Npk` and `Nsk` denote the size of public and private keys, respectively,
 used in the AKE.
 
-# Offline Registration {#offline-phase}
+- Deterministic Key Generation Functions:
+  - RecoverPublicKey(private_key): Recover the public key related to the input private key.
+  - DeriveKeyPair(seed): Derive a private and public key pair deterministically from the seed.
 
-Registration is executed between a client C and a
-server S. It is assumed S can identify C and the client can
-authenticate S during this registration phase. This is the only part
-in OPAQUE that requires an authenticated and confidential channel, either physical, out-of-band,
-PKI-based, etc. This section describes the registration flow, message encoding,
-and helper functions. Moreover, C has a key pair (client_private_key, client_public_key) for an AKE protocol
-which is suitable for use with OPAQUE; See {{online-phase}}. The private-public keys (client_private_key, client_public_key) may be randomly generated (using a cryptographically secure pseudorandom number generator) for the account or provided by the calling client.
-Clients MUST NOT use the same key pair (client_private_key, client_public_key) for two different accounts.
+Generally, nonces are of length `Nn` set to 32 bytes.
 
-## Setup Phase {#setup-phase}
+# Client Credential Storage {#client-credential-storage}
 
-In a setup phase, C chooses its password, and S chooses its own pair of private-public
-AKE keys (server_private_key, server_public_key) for use with the AKE, along with a Nh-byte oprf_seed. S can use
-the same pair of keys with multiple clients, and can opt to use multiple seeds (so long as they are
-kept consistent for each client). These steps can happen offline, i.e., before the registration phase.
+OPAQUE makes use of a structure `Envelope` to manage client credentials. This envelope holds information about its format and content for the client to obtain its authentication material.
 
-Once complete, the registration process proceeds as follows.
+OPAQUE allows applications to either provide custom client private and public keys for authentication or to generate them internally, making the application oblivious to the client's private key. Each public and private key value is an opaque byte string, specific to the AKE protocol in which OPAQUE is instantiated.
 
-## Credential Registration
+These two options are defined as the `Internal Key` and `External Key` modes.  See {{envelope-modes}} for their specifications.
```

Comment:
>OPAQUE allows applications to either provide custom client private and public keys for authentication or to generate them internally, making the application oblivious to the client's private key. Each public and private key value is an opaque byte string, specific to the AKE protocol in which OPAQUE is instantiated.
>
>These two options are defined as the Internal Key and External Key modes. See {{envelope-modes}} for their specifications.

Could we maybe be even more explicit to state that the the `EnvelopeMode` is external when a separate application(s) provide custom client private and public keys for authentication? 

>or to generate them internally

And then a secondary statement which states the complimentary where the `EnvelopeMode` is internal when it is not the case when a separate application(s) provide custom client private and public keys for authentication. 

Apologies if I'm falling out of the sky into this PR.  I've been following OPAQUE and TLS OPAQUE loosely for a bit.

Reply:
Hello 👋

The key can be provided by the same application or the wrapper protocol. A usage example is when the same application needs the private key for some other use (with the internal mode, the private key never leaves the AKE). So it's not just for use with external applications.

Reply:
Does this answer your question? Happy to discuss if you don't agree :)

Reply:
>A usage example is when the same application needs the private key for some other use (with the internal mode, the private key never leaves the AKE).

That is a lot clearer thank you @bytemare  !

Reply:
But you are right, this is a totally valid use case and we could include that example. Thanks!

Reply:
Do you think that the following phrase

> OPAQUE allows applications to either provide custom client private and public keys for authentication

covers your following use case not explicitly enough?

> when a separate application(s) provide custom client private and public keys for authentication

Even if provided by an external application, it is still the wrapper application that implements or integrates OPAQUE that will import the keys.